java/algorithm/src/main/java/org/apache/arrow/algorithm/sort/FixedWidthInPlaceVectorSorter.java

```diff
@@ -85,8 +91,54 @@ private void quickSort() {
     }
   }
 
+  /**
+   *  Select the pivot as the median of 3 samples.
+   */
+  void choosePivot(int low, int high) {
+    if (high - low < PIVOT_SELECTION_THRESHOLD) {
```

Comment:
Note that instead of skipping pivot selection, the threshold should probably trigger the use of insertion sort on small subarrays. Usually, quicksort implementations switch to insertion sort for arrays smaller than 10-20 elements.

Reply:
@pitrou Thanks a lot for the good suggestion. I agree with you that it is an optimization that will definitely improve the sort performance, and we really need it in our code base.

I am not sure if we should implement it in this particular PR, or in a separate one in the future.
Maybe insertion sort is a separate feature (I don't know), as it has some distinct advantages compared with quick sort:

1. It is extremely efficient for small vectors (as you have pointed out)
2. For an already sorted vector, its time complexity is O(n) (vs. O(nlogn) for quick sort).

Would you please give your suggestion?

Reply:
As you prefer of course, but since you are improving quick sort performance, it's probably a good idea to apply all the usual tricks at once :-)

Reply:
I see. Thanks a lot for your feedback. 
I will follow your advice, and provide the implementation of insertion sort in this PR :-)

Reply:
For the avoidance of ambiguity, I only meant use insertion sort for small subarrays in quicksort. Not expose it as an API (which I don't think is useful).

Reply:
I see. Thank you so much for the additional suggestion.

Reply:
@pitrou I have implemented the insertion sort algorithm, and made it package private. Would you please give another look?

Reply:
Sounds good on the principle, but I'll let a Java developer comment on the details.

Reply:
@pitrou Thanks a lot for your effort.